packit_service/worker/handlers/copr.py

```diff
@@ -479,3 +496,157 @@ def handle_testing_farm(self):
+        (with `commit` trigger and same branch configured as the target PR branch).
+        """
+        base_build_job = None
+
+        for job in self.package_config.get_job_views():
+            if (
+                job.type in (JobType.copr_build, JobType.build)
+                and job.trigger == JobConfigTriggerType.commit
+                and (
+                    (
+                        job.branch
+                        and job.branch
+                        == self.copr_build_helper.pull_request_object.target_branch
+                    )
+                    or (
+                        not job.branch
+                        and self.project.default_branch
+                        == self.copr_build_helper.pull_request_object.target_branch
+                    )
+                )
+            ):
+                base_build_job = job
+                break
+
+        return base_build_job
+
+    def get_base_srpm_model(
+        self, base_build_job: JobConfig
+    ) -> Optional[SRPMBuildModel]:
+        """
+        Get the SRPM build model of the latest successful Copr build
+        for the given job config.
+        """
+        base_build_project_name = (
+            self.copr_build_helper.job_project_for_commit_job_config(base_build_job)
+        )
+        base_build_owner = self.copr_build_helper.job_owner_for_job_config(
+            base_build_job
+        )
+        target_branch_commit = (
+            self.copr_build_helper.pull_request_object.target_branch_head_commit
+        )
+
+        logger.debug(
+            f"Searching for base build for {target_branch_commit} commit "
+            f"in {base_build_owner}/{base_build_project_name} Copr project in our DB. "
+        )
+
+        builds = CoprBuildTargetModel.get_all_by(
+            commit_sha=target_branch_commit,
+            project_name=base_build_project_name,
+            owner=base_build_owner,
+            target="fedora-rawhide-x86_64",
+            status=BuildStatus.success,
+        )
+        if not builds:
+            logger.debug("No matching base build found in our DB.")
+            return None
+
+        return list(builds)[0].get_srpm_build()
+
+    @staticmethod
+    def download_srpms(
+        directory: str,
+        base_srpm_model: SRPMBuildModel,
+        srpm_model: SRPMBuildModel,
+    ) -> Optional[tuple[Path, Path]]:
+        base_srpm_path = Path(directory).joinpath(basename(base_srpm_model.url))
+        if not download_file(base_srpm_model.url, base_srpm_path):
+            logger.info(
+                f"Downloading of base SRPM {base_srpm_model.url} was not successful."
+            )
+            return None
+
+        srpm_path = Path(directory).joinpath(basename(srpm_model.url))
+        if not download_file(srpm_model.url, srpm_path):
+            logger.info(f"Downloading of SRPM {srpm_model.url} was not successful.")
+            return None
```

Comment:
Nitpick - this is almost completely duplicated code, it could be avoided by e.g. using a nested function.